### PR TITLE
LPS-170425 | Master 

### PR DESCRIPTION
**Pull request by MirnaGama**
_Avoid updating password and sending email notification in case of AuthenticationException: [LDAP: Error code 49 - INVALID_CREDENTIALS]_

**Bug found:** A customer reported a behavior where Users receive an internal server error upon password change when connected to a LDAP server.

**Actual behavior:** Internal Server Error occurs on the UI, but still the "Your Password Has Been Changed" mail is sent.
**Expected Result:** A more User friendly error could be shown on the UI, so the User knows why it happened, and also the mail should not be sent.

**Commits included:**
- LPS-170425 Avoid importing and updating user in case of authentication failure in LDAP
- LPS-170425 If user is imported by LDAP then avoid authenticate omniadmin by default
- LPS-170425 Validate auth before update user and send email
- LPS-170425 Display LDAP Authentication Error in UI
- LPS-170425 SF
- LPS-170425 Create new exception for LDAP Invalid Credentials
- LPS-170425 Update language keys
- LPS-170425 SF
- LPS-170425 Update the user info and throw exception before send email notification
- LPS-170425 SF

Previously sent to: https://github.com/liferay-usm/liferay-portal/pull/915

### CODE REVIEW

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java [hunk 1]
```diff
@@ -375,48 +386,42 @@ private int _authenticate(
...
+
+						if (errorMessage.contains(
+								errorPasswordExpiredKeyword)) {
 
-					if (errorMessage.contains(errorPasswordExpiredKeyword)) {
-						throw new PasswordExpiredException();
+							throw new PasswordExpiredException();
+						}
 					}
 				}
```
Comment (oliverkecskemety) [inline modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:423]: This part was previously not inside the `if (!ldapAuthResult.isAuthenticated())` condition. As I see the `errorMessage` is only getting set after `isAuthenticated` is set to false but have you validated that it is true and no other scenario can happen?

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java [hunk 2]
```diff
@@ -375,48 +386,42 @@ private int _authenticate(
 			LDAPAuthResult ldapAuthResult = _authenticate(
 				safeLdapContext, companyId, attributes, fullUserDN, password);
 
-			// Get user or create fromUnsafe LDAP
-
 			if (!ldapAuthResult.isAuthenticated()) {
 				password = null;
```
Comment (oliverkecskemety) [inline modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:390]: This line will never have an effect as we will return `FAILURE` before reaching `_ldapUserImporter.importUser(ldapServerId, companyId, safeLdapContext, attributes, password);`

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java [hunk 3]
```diff
@@ -429,6 +434,11 @@ private int _authenticate(
 				return FAILURE;
 			}
 
+			// Get user or create fromUnsafe LDAP
+
+			User user = _ldapUserImporter.importUser(
+				ldapServerId, companyId, safeLdapContext, attributes, password);
```
Comment (oliverkecskemety) [inline modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:440]: Do you think we will never require to import a user from LDAP if the authentication fails?

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java [hunk 4]
```diff
@@ -538,8 +559,12 @@ private int _authenticate(
...
 		}
 
-		return _authenticateRequired(
-			companyId, userId, emailAddress, screenName, true, FAILURE);
+		if (!userImportedByLDAP) {
+			return _authenticateRequired(
+				companyId, userId, emailAddress, screenName, true, FAILURE);
+		}
+
+		return FAILURE;
```
Comment (oliverkecskemety) [inline modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:567]: Can you help me what your changes do here? It clearly changes the result fundamentally but I don't really understand why.

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java [hunk 5]
```diff
@@ -512,9 +523,19 @@ private int _authenticate(
...
 				ldapServerConfigurations) {
 
+			Binding ldapUser = _portalLDAP.getUser(
+				ldapServerConfiguration.ldapServerId(), companyId, screenName,
+				emailAddress);
+
+			if (ldapUser != null) {
+				userImportedByLDAP = true;
+			}
+
```
Comment (oliverkecskemety) [inline modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:538]: Same question as below. I couldn't figure out why is this required.

File: portal-impl/src/com/liferay/portal/action/UpdatePasswordAction.java [hunk 6]
```diff
@@ -302,9 +305,31 @@ else if (authType.equals(CompanyConstants.AUTH_TYPE_ID)) {
...
+		}
+		finally {
+			PwdToolkitUtilThreadLocal.setValidate(previousValidate);
+		}
+
+		if (ticket != null) {
+			TicketLocalServiceUtil.deleteTicket(ticket);
+
+			UserLocalServiceUtil.updatePasswordReset(userId, false);
+		}
```
Comment (oliverkecskemety) [inline portal-impl/src/com/liferay/portal/action/UpdatePasswordAction.java:332]: Ticket has already been deleted to this point by Line 284-292 even if the password change was unsuccessful. You might need to set `UserLocalServiceUtil.updatePasswordReset(userId, false)` for successful login but you should set it back if it failed.

File: portal-impl/src/com/liferay/portal/action/UpdatePasswordAction.java [hunk 7]
```diff
@@ -302,9 +305,31 @@ else if (authType.equals(CompanyConstants.AUTH_TYPE_ID)) {
...
-		AuthenticatedSessionManagerUtil.login(
-			httpServletRequest, httpServletResponse, login, password1, false,
-			null);
+		try {
+			boolean currentValidate = isValidatePassword(httpServletRequest);
+
+			PwdToolkitUtilThreadLocal.setValidate(currentValidate);
+
+			User userUpdated = UserLocalServiceUtil.updatePassword(
+				userId, password1, password2, passwordReset);
```
Comment (oliverkecskemety) [inline portal-impl/src/com/liferay/portal/action/UpdatePasswordAction.java:314]: For e-mail sending the already existing `user` object would be sufficient enough in my opinion.

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java [hunk 8]
```diff
@@ -4878,10 +4973,6 @@ public User updatePassword(
 
 		if (!silentUpdate) {
 			user.setPasswordModified(false);
-
-			sendPasswordNotification(
-				user, user.getCompanyId(), password1, null, null, null, null,
-				null, serviceContext);
```
Comment (oliverkecskemety) [inline portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java:4884]: Moving the `sendPasswordNotification` out from here will require third parties to reimplement their solutions I guess, so it would have quite a meaningful impact. I would avoid that if possible.

**Review by oliverkecskemety — CHANGES_REQUESTED** [review]
Hi @MirnaGama,

Thanks for the PR! Please see my comments.

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java (hunk 8 above)
Comment (MirnaGama) [reply portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java:4884]: I understand, but it's hard to get around the problem in the current way the code is implemented. It validates the login along with sending the email, so I thought of splitting the whole method into two.

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java (hunk 4 above)
Comment (MirnaGama) [reply modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:567]: I noticed that there is a odd behavior on this block since the ldap server authentication fails and even then the user is logged successfully.

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java (hunk 5 above)
Comment (MirnaGama) [reply modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:538]: This block checks if the user was imported by LDAP, so that automatic login does not happen.

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java (hunk 1 above)
Comment (MirnaGama) [reply modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/authenticator/LDAPAuth.java:423]: Yes, I just reduced the amount of verifications as I noticed that this `!ldapAuthResult.isAuthenticated()` condition was being called several times sequentially in a small time interval.

**Comment by MirnaGama** [comment]
New PR: https://github.com/liferay-appsec/liferay-portal/pull/977

**Comment by liferay-appsec** [comment]
Won't fix. https://issues.liferay.com/browse/LPS-170425
